### PR TITLE
Deprecate `weights_save_path` from the Trainer constructor 

### PR DESCRIPTION
**Pull request by daniellepintz**
## What does this PR do?
Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/11768

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by krshrimali — APPROVED** [review]
LGTM! 🚀

**Review by ananthsub — APPROVED** [review]

File: tests/deprecated_api/test_remove_1-8.py [hunk 1]
```diff
@@ -572,3 +572,12 @@ def agg_and_log_metrics(self, metrics, step):
         Trainer(logger=[logger, logger3])
     # Should have no deprecation warning
     Trainer(logger=[logger2, logger3])
+
+
+def test_v1_8_0_weights_save_path(tmpdir):
```
Comment (kaushikb11) [inline tests/deprecated_api/test_remove_1-8.py:577]:
```suggestion
def test_v1_8_0_weights_save_path():
```

File: tests/deprecated_api/test_remove_1-8.py [hunk 2]
```diff
@@ -572,3 +572,12 @@ def agg_and_log_metrics(self, metrics, step):
         Trainer(logger=[logger, logger3])
     # Should have no deprecation warning
     Trainer(logger=[logger2, logger3])
+
+
+def test_v1_8_0_weights_save_path():
```
Comment (kaushikb11) [inline tests/deprecated_api/test_remove_1-8.py:577]:
```suggestion
def test_v1_8_0_weights_save_path(tmpdir):
```

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -2206,6 +2211,9 @@ def weights_save_path(self) -> str:
         """
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:2211]: if someone acess this property, should a warning be raised here too?

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 4]
```diff
@@ -53,6 +53,13 @@ def on_trainer_init(
     ):
         # init folder paths for checkpoint + weights save callbacks
         self.trainer._default_root_dir = default_root_dir or os.getcwd()
+        if weights_save_path:
+            rank_zero_deprecation(
+                "Setting `Trainer(weights_save_path=)` is deprecated in v1.6 and will be removed in v1.8."
+                " Please pass ``dirpath`` directly to the"
+                " :class:`~pytorch_lightning.callbacks.model_checkpoint.ModelCheckpoint` callback."
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/callback_connector.py:60]:
```suggestion
                "Setting `Trainer(weights_save_path=)` has been deprecated in v1.6 and will be"
                " removed in v1.8. Please pass ``dirpath`` directly to the `ModelCheckpoint` callback"
```

File: docs/source/common/trainer.rst [hunk 5]
```diff
@@ -1582,6 +1582,12 @@ Can specify as float or int.
 weights_save_path
 ^^^^^^^^^^^^^^^^^
 
+
+.. warning:: `weights_save_path` is deprecated in v1.6 and will be removed in v1.8. Please pass
```
Comment (rohitgr7) [inline docs/source/common/trainer.rst:1586]:
```suggestion
.. warning:: `weights_save_path` has been deprecated in v1.6 and will be removed in v1.8. Please pass
```

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -2206,6 +2211,9 @@ def weights_save_path(self) -> str:
...
         The default root location to save weights (checkpoints), e.g., when the
         :class:`~pytorch_lightning.callbacks.model_checkpoint.ModelCheckpoint` does not define a file path.
+
+        .. deprecated:: v1.6
+            `Trainer.weights_save_path` is deprecated in v1.6 and will be removed in v1.8.
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:2216]:
```suggestion
            `Trainer.weights_save_path` has been deprecated in v1.6 and will be removed in v1.8.
```

File: tests/deprecated_api/test_remove_1-8.py [hunk 7]
```diff
@@ -572,3 +572,12 @@ def agg_and_log_metrics(self, metrics, step):
         Trainer(logger=[logger, logger3])
     # Should have no deprecation warning
     Trainer(logger=[logger2, logger3])
+
+
+def test_v1_8_0_weights_save_path(tmpdir):
+    with pytest.deprecated_call(
+        match=r"Setting `Trainer\(weights_save_path=\)` is deprecated in v1.6 and will be removed in v1.8."
```
Comment (rohitgr7) [inline tests/deprecated_api/test_remove_1-8.py:579]:
```suggestion
        match=r"Setting `Trainer\(weights_save_path=\)` has been deprecated in v1.6"
```

File: tests/loggers/test_all.py [hunk 8]
```diff
@@ -216,7 +216,10 @@ def name(self):
     save_dir = tmpdir / "logs"
     weights_save_path = tmpdir / "weights"
     logger = TestLogger(**_get_logger_args(TestLogger, save_dir))
-    trainer = Trainer(**trainer_args, logger=logger, weights_save_path=weights_save_path)
+    with pytest.deprecated_call(
+        match=r"Setting `Trainer\(weights_save_path=\)` is deprecated in v1.6 and will be removed in v1.8."
```
Comment (rohitgr7) [inline tests/loggers/test_all.py:220]:
```suggestion
        match=r"Setting `Trainer\(weights_save_path=\)` has been deprecated in v1.6"
```

File: tests/loggers/test_all.py [hunk 9]
```diff
@@ -225,7 +228,10 @@ def name(self):
 
     # no logger given
     weights_save_path = tmpdir / "weights"
-    trainer = Trainer(**trainer_args, logger=False, weights_save_path=weights_save_path)
+    with pytest.deprecated_call(
+        match=r"Setting `Trainer\(weights_save_path=\)` is deprecated in v1.6 and will be removed in v1.8."
```
Comment (rohitgr7) [inline tests/loggers/test_all.py:232]:
```suggestion
        match=r"Setting `Trainer\(weights_save_path=\)` has been deprecated in v1.6"
```

**Review by rohitgr7 — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]
Once Rohit's comments are addressed, LGTM!

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:2211]: Unfortunately we can't do that, since we still need to use `trainer.weights_save_path` in the codebase until it is removed, like here: https://github.com/PyTorchLightning/pytorch-lightning/blob/a0655611de460f1659b46150cda256d2e3fa974e/pytorch_lightning/callbacks/model_checkpoint.py#L579-L592

I think it is sufficient to warn if someone passes `weights_save_path` to the Trainer as an argument.

**Comment by daniellepintz** [comment]
Could someone please rerun TPU tests and enable auto merge?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 10]
```diff
@@ -55,11 +55,12 @@ def __init__(self, trainer: "pl.Trainer", resume_from_checkpoint: Optional[_PATH
 
     @property
     def _hpc_resume_path(self) -> Optional[str]:
-        weights_save_path = self.trainer.weights_save_path
-        fs = get_filesystem(weights_save_path)
-        if not fs.isdir(weights_save_path):
+        # in v1.8 set this equal to self.trainer.default_root_dir
+        dir_path_hpc = self.trainer.weights_save_path
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:59]: Can you also update all the references of `trainer.weights_save_path` in the tests (even though you don't raise a warning)?

File: tests/deprecated_api/test_remove_1-8.py [hunk 11]
```diff
@@ -608,3 +608,10 @@ def on_pretrain_routine_end(self, trainer, pl_module):
         match="The `Callback.on_pretrain_routine_end` hook has been deprecated in v1.6" " and will be removed in v1.8"
     ):
         trainer.fit(model)
+
+
+def test_v1_8_0_weights_save_path(tmpdir):
+    with pytest.deprecated_call(match=r"Setting `Trainer\(weights_save_path=\)` has been deprecated in v1.6"):
+        _ = Trainer(
+            weights_save_path=tmpdir,
+        )
```
Comment (awaelchli) [inline tests/deprecated_api/test_remove_1-8.py:617]:
```suggestion
        _ = Trainer(weights_save_path=tmpdir)
```

Comment (awaelchli) [reply tests/deprecated_api/test_remove_1-8.py:617]: this test should also confirm that the modelcheckpoint callback dir path is still set to weight save path so we can refactor the other tests, right?

Comment (daniellepintz) [reply tests/deprecated_api/test_remove_1-8.py]: I believe this is done in `test_all` below?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (daniellepintz) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:59]: > so we aren't emitting a warning when someone calls the getter for weights_save_path? Because we want to avoid causing a warning when using it internally?

that's correct. 

> Can you also update all the references of trainer.weights_save_path in the tests (even though you don't raise a warning)?

The only references of trainer.weights_save_path in the tests are in `loggers/test_all.py` which I believe are all necessary to test that weights_save_path is still working as intended until it is removed.

[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:59]: @daniellepintz we could define a property & setter for weights save path on the trainer which emit a deprecation warning.

And usages within the framework can rely on a private attribute which backs the property's implementation, so no deprecation warnings are emitted then.

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:59]: Yes agreed, if it's possible I would definitely do that. The benefits of removing all deprecated references and re-writing the tests in the deprecation PR:
- The follow up removal in v1.8 is dead simple and can be done by anyone (good first issue)
- We can be certain already now that the removal will not need additional fixing in 1.8, because we adjust the tests now.

I think we want to follow this as a general rule for all removals.

**Review by awaelchli — APPROVED** [review]

File: tests/deprecated_api/test_remove_1-8.py (hunk 11 above)
Comment (awaelchli) [reply tests/deprecated_api/test_remove_1-8.py]: what I'm trying to say is, we can make deprecation test here in test_remove_1-8.py where we do:

```py
trainer = Trainer(weight_save_path="a/b/c")
assert trainer.checkpoint_callback.dir_path == "a/b/c"
```

so that we can update other tests to use the new API / get rid of weight_save_path.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (daniellepintz) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:59]: > we should define a property & setter for weights save path

@ananthsub why a setter? we don't have one now?

[the remark this answers is not in the document]
Comment (daniellepintz) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: Ok, I added a `_weights_save_path_internal` property

File: tests/deprecated_api/test_remove_1-8.py (hunk 11 above)
Comment (daniellepintz) [reply tests/deprecated_api/test_remove_1-8.py]: @awaelchli I was under the impression that in the `test_remove_x-y.py` we generally want to test the deprecation warnings, and the functionality of the deprecated behavior should be tested in the same tests that we have been using  prior to deprecation. Following this logic, shouldn't we keep this test here which tests the functionality of weights_save_path? Is my understanding wrong of what we use the `test_remove_x-y.py` for?

https://github.com/PyTorchLightning/pytorch-lightning/blob/29d5afbda6ad64e59241950d0b94b19564c2710e/tests/loggers/test_all.py#L223

File: pytorch_lightning/trainer/configuration_validator.py [hunk 12]
```diff
@@ -361,5 +361,5 @@ def _check_deprecated_callback_hooks(trainer: "pl.Trainer") -> None:
             if is_overridden(method_name=hook, instance=callback):
                 rank_zero_deprecation(
                     f"The `Callback.{hook}` hook has been deprecated in v1.6 and"
-                    f" will be removed in v1.8. Please use `Callback.on_fit_start` instead."
+                    " will be removed in v1.8. Please use `Callback.on_fit_start` instead."
```
Comment (daniellepintz) [inline pytorch_lightning/trainer/configuration_validator.py:364]: this is a small fix from a previous PR

File: tests/deprecated_api/test_remove_1-8.py [hunk 13]
```diff
@@ -588,7 +588,7 @@ def on_pretrain_routine_start(self, trainer, pl_module):
         default_root_dir=tmpdir,
     )
     with pytest.deprecated_call(
-        match="The `Callback.on_pretrain_routine_start` hook has been deprecated in v1.6" " and will be removed in v1.8"
+        match="The `Callback.on_pretrain_routine_start` hook has been deprecated in v1.6 and will be removed in v1.8"
```
Comment (daniellepintz) [inline tests/deprecated_api/test_remove_1-8.py:591]: small fix from a previous PR

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 14]
```diff
@@ -573,10 +573,11 @@ def __resolve_ckpt_dir(self, trainer: "pl.Trainer") -> None:
         if self.dirpath is not None:
             return  # short circuit
 
+        # TODO: Remove weights_save_path logic here in v1.8
         if trainer.logger is not None:
-            if trainer.weights_save_path != trainer.default_root_dir:
+            if trainer._weights_save_path_internal != trainer.default_root_dir:
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:578]:
```suggestion
            if trainer._weights_save_path != trainer.default_root_dir:
```
if you really mean it as internal, lets make is private or protected

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 14 above)
Comment (daniellepintz) [reply pytorch_lightning/callbacks/model_checkpoint.py:578]: I'm confused - it is private already?

Comment (daniellepintz) [reply pytorch_lightning/callbacks/model_checkpoint.py:580]: oh I think I know what you mean. I don't want to just change `weights_save_path` to `_weights_save_path` because we already have `_weights_save_path` which does something else:
https://github.com/PyTorchLightning/pytorch-lightning/blob/dc4c3171fc0618fb463a846ccbde2e558902303b/pytorch_lightning/trainer/trainer.py#L2207-L2215

**Comment by daniellepintz** [comment]
Could someone please rerun TPU tests and enable auto merge?
